File: top_k_freq_elements.py
# Approach - since we care about k elements out of N we can use heaps
# Based on freq push to min heap and pop smaller number up until size k

# Time - O(N) + O(N log K) ~= O(N log K)

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        # min heap solution
        min_heap = []
        count = collections.Counter(nums) # O(N)
        
        for num, freq in count.items(): # O(N log K)
            heappush(min_heap, (freq, num))
            
            if len(min_heap) > k:
                heappop(min_heap)
              
        result = []
                
        for freq, num in min_heap:
            result.append(num)
            
        return result
        
        
        
        
        
# Brute Force
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        dict = Counter(nums) # O(N)

        k_freq = dict.most_common(k) 
        for val in k_freq:
            result.append(val[0])
            
        return result
                

File: TopKFreqElements.java
//Time: O(N) | Space: O(N+K)

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;
        for(int i=0;i<nums.length;i++) {
            int freq = map.getOrDefault(nums[i], 0)+1;
            map.put(nums[i],freq);
            max = Math.max(max, freq);
            min = Math.min(min, freq);
        }

        Map<Integer, List<Integer>> freqMap = new HashMap<>();
        for(int key: map.keySet()) {
            int freq = map.get(key);
            if(!freqMap.containsKey(freq)){
                freqMap.put(freq, new ArrayList<>());
            }
            freqMap.get(freq).add(key);
        }
        int[] result = new int[k];
        int count = 0;
        for(int i=max;i>=min && count < k;i--){
            if(freqMap.containsKey(i)) {
                List<Integer> li = freqMap.get(i);
                for(int item: li) {
                    result[count] = item;
                    count++;
                }
            }
        }
        return result;
    }
}
//Heap Soln
//Time: O(Nlogk) | Space: O(N)
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int i=0;i<nums.length;i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0)+1);
        }
        PriorityQueue<Integer> pq = new PriorityQueue<>((a,b) -> map.get(a) - map.get(b));
        for(int ele: map.keySet()) {
            pq.add(ele);
            if(pq.size() > k)
                pq.poll();
        }

        int[] result = new int[k];
        for(int i=0;i<k;i++) {
            result[i] = pq.poll();
        }
        return result;
    }
}